presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -79,31 +90,163 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
-            Expression predicate = node.getPredicate();
-            if (!SymbolsExtractor.extractUnique(predicate).containsAll(correlation)) {
+            Optional<DecorrelationResult> childDecorrelationResultOptional = Optional.of(new DecorrelationResult(
+                    node.getSource(),
+                    ImmutableSet.of(),
+                    ImmutableList.of(),
+                    ImmutableMultimap.of(),
+                    false));
+
+            // try to decorrelate filters down the tree
+            if (containsCorrelation(node.getSource(), correlation)) {
+                childDecorrelationResultOptional = lookup.resolve(node.getSource()).accept(this, null);
+            }
+
+            if (!childDecorrelationResultOptional.isPresent()) {
                 return Optional.empty();
             }
 
+            Expression predicate = node.getPredicate();
             Map<Boolean, List<Expression>> predicates = ExpressionUtils.extractConjuncts(predicate).stream()
                     .collect(Collectors.partitioningBy(isUsingPredicate(correlation)));
             List<Expression> correlatedPredicates = ImmutableList.copyOf(predicates.get(true));
             List<Expression> uncorrelatedPredicates = ImmutableList.copyOf(predicates.get(false));
 
+            DecorrelationResult childDecorrelationResult = childDecorrelationResultOptional.get();
             FilterNode newFilterNode = new FilterNode(
                     idAllocator.getNextId(),
-                    node.getSource(),
+                    childDecorrelationResult.node,
                     ExpressionUtils.combineConjuncts(uncorrelatedPredicates));
 
+            Set<Symbol> symbolsToPropagate = Sets.difference(SymbolsExtractor.extractUnique(correlatedPredicates), ImmutableSet.copyOf(correlation));
             return Optional.of(new DecorrelationResult(
                     newFilterNode,
-                    Sets.difference(SymbolsExtractor.extractUnique(correlatedPredicates), ImmutableSet.copyOf(correlation)),
-                    correlatedPredicates));
+                    Sets.union(childDecorrelationResult.symbolsToPropagate, symbolsToPropagate),
+                    ImmutableList.<Expression>builder()
+                            .addAll(childDecorrelationResult.correlatedPredicates)
+                            .addAll(correlatedPredicates).build(),
+                    ImmutableMultimap.<Symbol, Symbol>builder()
+                            .putAll(childDecorrelationResult.correlatedSymbolsMapping)
+                            .putAll(extractCorrelatedSymbolsMapping(correlatedPredicates))
+                            .build(),
+                    childDecorrelationResult.atMostSingleRow));
+        }
+
+        @Override
+        public Optional<DecorrelationResult> visitLimit(LimitNode node, Void context)
+        {
+            Optional<DecorrelationResult> childDecorrelationResultOptional = lookup.resolve(node.getSource()).accept(this, null);
```

Comment:
`visitFilter` has `if (containsCorrelation(node.getSource(), correlation))` check, but this method (`visitLimit`) does not; why is the difference?

Reply:
`FilterNodes` are "special" nodes form a correlated subquery perspective because they provide binding (as predicates) between subquery symbols and correlated (outer) symbols. `FilterNodes` are essentially a starting point for decorrelation process.

Only filter nodes can provide required info (e.g: mapping from correlated symbol to subquery symbol) for other nodes to be decorrelated. Because of that filter nodes are the only nodes that prevent visitor from going deeper (they don't require further information to decorrelate)

I hope that answers your question